### PR TITLE
Save user xp to database

### PR DESCRIPTION
**Pull request by turtleman270**
Able to save xp to a firebase table
Able to load xp from that table

I did some refactoring to make saving/loading from a db easier
Everything is still functionally the same as before afaik

UI changes
 - added another column with save/load user buttons

### CODE REVIEW

**Comment by turtleman270** [comment]
I wasn't able to create a branch on your repo so I had to fork it to make a PR

File: src/App.js [hunk 1]
```diff
@@ -87,39 +90,46 @@ const App = () => {
 	}
 
 	//Gameplay
-	const levelMarks = [1, 2, 4, 8, 12, 18, 24, 36, 48, 60, 75, 100, 125, 150, 175, 200, 225, 260, 300, 340, 375, 425, 500]
```
Comment (turtleman270) [inline src/App.js:90]: This seemed pretty arbitrary to me, and didn't make things super easy when trying to find the current level based on xp

File: src/App.js [hunk 2]
```diff
@@ -181,23 +191,63 @@ const App = () => {
...
+											sacrificeXp: sacrifice.exp}
+		const response = fetch("https://skilling-59c90.firebaseio.com/anthony.json", {
+	    method: 'PUT',
+	    body: JSON.stringify(userData)
+	  });
+	}
+
 	useEffect(() => {
 		//Unlock Tool: Saw
-		if (woodcutting.exp === 9 && tools.some(e => e.name !== 'hammer')) {
```
Comment (turtleman270) [inline src/App.js:186]: This wasn't working as you intended

In tools, is there some/any value where name != hammer? (that's nearly always true)

**Comment by turtleman270** [comment]
Please ask questions before merging this :)

File: src/App.js [hunk 3]
```diff
@@ -87,39 +90,46 @@ const App = () => {
...
-		}
+	const levelFormula = (exp) => {
+		return Math.floor(Math.sqrt(exp)*1.2);
+	}
+
+	const updateLevel = (skillString, skillName, skillSetter, exp) => {
+		exp = isNaN(exp) ? 0 : exp
+		const level = levelFormula(exp)
+		skillSetter({ exp: exp, level: level })
+	}
```
Comment (theRJMurray) [inline src/App.js:101]: the parameter skillString in the updateLevel function doesn't need to be there. Am I correct?

File: src/App.js [hunk 4]
```diff
@@ -232,9 +282,14 @@ const App = () => {
 					{tools.some(e => e.name === 'lighter') && <Skill takeAction={craftCampfire} skill={makeCampfire} skillName={"Campfire Crafting"} skillAction={"Craft Campfire"} />}
 					{tools.some(e => e.name === 'lighter') && <Skill takeAction={sacrificeFish} skill={sacrifice} skillName={"Sacrifice"} skillAction={"Sacrifice Fish"} />}
 				</div>
+				<div className="craft">
+					{<Skill takeAction={login} skill={toolMaking} skillName={"Load User"} skillAction={"Anthony"} />}
+					{<Skill takeAction={save} skill={makeCampfire} skillName={"Save User"} skillAction={"Anthony"} />}
+				</div>
+
```
Comment (theRJMurray) [inline src/App.js:289]: I love how you made the buttons have the text 'Anthony' hahaha

File: src/App.js (hunk 2 above)
Comment (theRJMurray) [reply src/App.js:186]: Yeah I noticed it was buggy before. Seems like you've fixed it though.

File: src/App.js [hunk 5]
```diff
@@ -181,23 +191,63 @@ const App = () => {
...
+											cookXp: cooking.exp,
+											prospectXp: prospecting.exp,
+											toolXp: toolMaking.exp,
+											sacrificeXp: sacrifice.exp}
+		const response = fetch("https://skilling-59c90.firebaseio.com/anthony.json", {
+	    method: 'PUT',
+	    body: JSON.stringify(userData)
+	  });
+	}
+
```
Comment (theRJMurray) [inline src/App.js:232]: doesn't update the inventory items at the top of the UI. I could add that?

File: src/App.js [hunk 6]
```diff
@@ -87,39 +90,46 @@ const App = () => {
...
+		skillSetter({ exp: exp, level: level })
+	}
+	const updateTotalLevel = () => {
+		setTotalLevel(allSkills.reduce((a, b) => a + b.level, 0))
+	}
+
+	const gainXp = (skillString, skillName, skillSetter) => {
+		const exp = skillName.exp + 1
+		updateLevel(skillString, skillName, skillSetter, exp)
 	}
```
Comment (theRJMurray) [inline src/App.js:109]: `	const baseStats = {
name: '',		
exp: 0,
		level: 0
	}`

Were you thinking of plugging the skillString in like this?

File: src/App.js [hunk 7]
```diff
@@ -181,23 +191,63 @@ const App = () => {
...
+					updateLevel("cook", cooking, setCooking, result.cookXp)
+					updateLevel("prospect", prospecting, setProspecting, result.prospectXp)
+					updateLevel("tool", toolMaking, setToolMaking, result.toolXp)
+					updateLevel("sacrifice", sacrifice, setSacrifice, result.sacrificeXp)
+        },
+        (error) => {
+          this.setState({
+            isLoaded: true,
+            error
+          });
```
Comment (theRJMurray) [inline src/App.js:213]: `  this.setState({
            isLoaded: true,
            error
          });`

I'm getting an error when I try to plug in my own Firebase URL and call the login function by clicking Load User.

Also, not sure what this bit of code is supposed to do. Could you explain?

**Review by theRJMurray — COMMENTED** [review]
I was thinking of making a branch on your fork then doing a PR before we merge to the main repository. Not sure if good idea or not. What do you think?

**Comment by theRJMurray** [comment]
Love the new functions you've added to make things smoother:
`  const levelFormula = exp => {
    return Math.floor(Math.sqrt(exp) * 1.2);
  };`

`  const updateLevel = (skillString, skillName, skillSetter, exp) => {
    exp = isNaN(exp) ? 0 : exp;`

`  const updateTotalLevel = () => {
    setTotalLevel(allSkills.reduce((a, b) => a + b.level, 0));
  };`

`  const gainXp = (skillString, skillName, skillSetter) => {
    const exp = skillName.exp + 1;};`

I had no idea connecting to Firebase was this easy!

Awesome changes.

**Comment by turtleman270** [comment]
> I was thinking of making a branch on your fork then doing a PR before we merge to the main repository. Not sure if good idea or not. What do you think?

I don't think it'll make any difference either way

File: src/App.js (hunk 3 above)
Comment (turtleman270) [reply src/App.js:101]: Good catch

I was using skillString before to update an object that had all the levels

ex. 
allLevels = {
woodcutting: 49,
fishing: 12,
...
}

The goal was to sum all the levels to get the total level

File: src/App.js (hunk 6 above)
Comment (turtleman270) [reply src/App.js:109]: Kind of, I was wanting an easy way to iterate over all the different skills to get the total level

File: src/App.js (hunk 7 above)
Comment (turtleman270) [reply src/App.js:213]: I have no idea, it was a copy/paste from a http example.

Does your Firebase db have public read/write?

File: src/App.js (hunk 5 above)
Comment (turtleman270) [reply src/App.js:232]: Please do, it'd also be a good starting point for a bit of db exposure

File: src/App.js (hunk 4 above)
Comment (turtleman270) [reply src/App.js:289]: It's hardcoded to only work with the user "Anthony"

**Pull request merged**